### DISCUSSION TITLE
[Python] How to implement the equivalent of pandas `dropna()` with the Expression API?

### DISCUSSION THREAD

**Discussion by ToddSmall** [duckdb/duckdb#10059] [Q&A]
I'd like to implement a simplified version of [`pandas.DataFrame.dropna()`](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.dropna.html). The goal of my simplified version is to drop rows for which any of the fields is `NULL`. My sticking point is that I don't know how to test that a `ColumnExpression` is `NULL`. There are many examples of the usage of `ColumnExpression` in [`test_expression.py`](https://github.com/duckdb/duckdb/blob/a491470b039c54fe2f0adfe1d161b14c7fe64642/tools/pythonpkg/tests/fast/test_expression.py), but no examples of testing a column for `NULL`s.

_If_ I could test `ColumnExpressions` for `NULL`s, then I think the rest of my function would proceed as illustrated below:

```
import functools

import duckdb
import pandas as pd

# `True` means that the corresponding value is `NULL`
df_test = pd.DataFrame(
    {
        "a": [True, True, True, False],
        "b": [True, True, True, False],
        "c": [True, True, False, False],
        "d": [True, False, False, False],
    }
)

df_test = duckdb.df(df_test)

cols = [duckdb.ColumnExpression(c) for c in df_test.columns]

reduced_or = functools.reduce(lambda x, y: x.__or__(y), cols).alias("should_drop")

df_test.select(duckdb.StarExpression()).filter(~reduced_or)

# The output is one row: false, false, false, false
```

I'd be grateful for suggestions for how to test a `ColumnExpression` for `NULL`s and for any other suggestions for how to improve my code.

Thank you!

**Comment by cmdlineluser** [comment]
Disclaimer: I've not used the Expression API so this is open for correction.

I cannot find anything related to NULL.

It seems like you're looking for [COLUMNS](https://duckdb.org/docs/sql/expressions/star#columns-expression) - but I can't find that in the expression api either.

```python
df = pd.DataFrame({
    "a": [None, None, None, 1],
    "b": [None, None, None, 2],
    "c": [None, None, 3, 4],
    "d": [None, 5, 6, 7],
})

df = duckdb.df(df)

# ┌────────┬────────┬────────┬────────┐
# │   a    │   b    │   c    │   d    │
# │ double │ double │ double │ double │
# ├────────┼────────┼────────┼────────┤
# │   NULL │   NULL │   NULL │   NULL │
# │   NULL │   NULL │   NULL │    5.0 │
# │   NULL │   NULL │    3.0 │    6.0 │
# │    1.0 │    2.0 │    4.0 │    7.0 │
# └────────┴────────┴────────┴────────┘
```

```python
df.select("row_number() over() row_nr, *").filter("columns(*) is not null")

# ┌────────┬────────┬────────┬────────┬────────┐
# │ row_nr │   a    │   b    │   c    │   d    │
# │ int64  │ double │ double │ double │ double │
# ├────────┼────────┼────────┼────────┼────────┤
# │      4 │    1.0 │    2.0 │    4.0 │    7.0 │
# └────────┴────────┴────────┴────────┴────────┘
```

(I think the `is not null` may also be superfluous here?)

**Reply by ToddSmall** [reply]
Thank you for your answer, @cmdlineluser!

Is it documented how `filter()` handles multiple boolean columns? It seems to combine the boolean values from, in this case, the four columns generated with `COLUMNS(*) IS NOT NULL` with ~`OR`~`AND`.

While I was fiddling around with options, I also noticed that I can't use `COALESCE` in a `FunctionExpression` because, I think, `COALESCE` is an operator and not a function.

```
df = pd.DataFrame({
    "a": [None, None, None, 1],
    "b": [None, None, None, 2],
    "c": [None, None, 3, 4],
    "d": [None, 5, 6, 7],
})

df = duckdb.df(df)

# ┌────────┬────────┬────────┬────────┐
# │   a    │   b    │   c    │   d    │
# │ double │ double │ double │ double │
# ├────────┼────────┼────────┼────────┤
# │   NULL │   NULL │   NULL │   NULL │
# │   NULL │   NULL │   NULL │    5.0 │
# │   NULL │   NULL │    3.0 │    6.0 │
# │    1.0 │    2.0 │    4.0 │    7.0 │
# └────────┴────────┴────────┴────────┘

cols = [duckdb.ColumnExpression(c) for c in df_test.columns]
fn = duckdb.FunctionExpression("coalesce", *cols)

df.select(fn)

# CatalogException: Catalog Error: Scalar Function with name coalesce does not exist!
# Did you mean "cos"?
```

I wonder how one can use operators such as `COALESCE` in a `FunctionExpression`?

**Reply by cmdlineluser** [reply]
Is it not `AND`?

If it was `OR` - then rows 2, 3, and 4 would be returned instead?

As for documentation, I'm not sure if there is any - just in searching for more info I did find https://github.com/duckdb/duckdb/issues/2000

> We discussed this internally and concluded to close this, focusing our efforts on the PySpark-compatible API now.

https://duckdb.org/docs/api/python/spark_api.html

(It doesn't seem like COALESCE / NULL functionality has been implemented yet?)

**Reply by ToddSmall** [reply]
You're correct that it's `AND`. Thank you!

**Comment by ToddSmall** [comment]
Upon reflection, I realized that I could write something closer to what I was originally imagining by writing my own function to check for `NULL` values.

```
import functools

import duckdb
import pandas as pd
import pyarrow

def one_if_null(x):
    """Returns 1 if `x` is NULL, otherwise 0."""
    return pyarrow.compute.cast(pyarrow.compute.is_null(x), target_type=pyarrow.uint8())


duckdb.create_function(
    "one_if_null",
    one_if_null,
    None,
    duckdb.typing.UTINYINT,
    type="arrow",
    null_handling="special",
)
```

With this function, I can define a filter expression that counts the number of `NULL` values in each row:

```
filter_expr = functools.reduce(
    lambda x, y: x + y,
    [
        duckdb.FunctionExpression("one_if_null", duckdb.ColumnExpression(col))
        for col in df_test.columns
    ],
).alias("row_num_nulls")
```

If I want to keep only rows with (say) two or fewer `NULL`s, I can do thing following:

```
df_test = pd.DataFrame(
    {
        "a": [None, None, None, 1],
        "b": [None, None, None, 2],
        "c": [None, None, 3, 4],
        "d": [None, 5, 6, 7],
    }
)

df_test = duckdb.df(df_test)

# ┌────────┬────────┬────────┬────────┐
# │   a    │   b    │   c    │   d    │
# │ double │ double │ double │ double │
# ├────────┼────────┼────────┼────────┤
# │   NULL │   NULL │   NULL │   NULL │
# │   NULL │   NULL │   NULL │    5.0 │
# │   NULL │   NULL │    3.0 │    6.0 │
# │    1.0 │    2.0 │    4.0 │    7.0 │
# └────────┴────────┴────────┴────────┘

df_filtered = df_test.filter(filter_expr <= 2)

# ┌────────┬────────┬────────┬────────┐
# │   a    │   b    │   c    │   d    │
# │ double │ double │ double │ double │
# ├────────┼────────┼────────┼────────┤
# │   NULL │   NULL │    3.0 │    6.0 │
# │    1.0 │    2.0 │    4.0 │    7.0 │
# └────────┴────────┴────────┴────────┘
```

What I still don't understand how to do is to use the operators `IS NULL`/`IS NOT NULL` with a `ColumnExpression`.

**Reply by cmdlineluser** [reply]
From what I can tell, the null operators just haven't been exposed as of yet.

I'm very much an SQL novice so I may be mistaken, but my understanding is that "horizontal operations" like this require static column names.

```python
duckdb.sql("""
from df 
select (
    from (values (a), (b), (c), (d))
    select sum((columns(*) is null)::int) 
) null_count
""")

# ┌────────────┐
# │ null_count │
# │   int128   │
# ├────────────┤
# │          4 │
# │          3 │
# │          2 │
# │          0 │
# └────────────┘
```

```python
duckdb.sql("""
from df 
where (
    from (values (a), (b), (c), (d))
    select sum((columns(*) is null)::int) 
) <= 2
""")

# ┌────────┬────────┬────────┬────────┐
# │   a    │   b    │   c    │   d    │
# │ double │ double │ double │ double │
# ├────────┼────────┼────────┼────────┤
# │   NULL │   NULL │    3.0 │    6.0 │
# │    1.0 │    2.0 │    4.0 │    7.0 │
# └────────┴────────┴────────┴────────┘
```

It doesn't seem like this type of query can be built using ColumnExpression.

**Reply by ToddSmall** [reply]
Your `from (values ...) select sum(...)` is very clever! I would not have guessed that it would work without quoting a, b, c, d in the values statement. (The query still works fine quoting a, b, c, d.)

My idea for building this query from `ColumnExpression`s (plural) is to take advantage of the fact that `ColumnExpression`s can be combined, notably summed. For example:

```
df_test.select(duckdb.ColumnExpression("c") + duckdb.ColumnExpression("d"))

# ┌─────────┐
# │ (c + d) │
# │ double  │
# ├─────────┤
# │    NULL │
# │    NULL │
# │     9.0 │
# │    11.0 │
# └─────────┘
```

One can combine `ColumnExpression`s using `functools.reduce` and a relation's `.columns` attribute (although for a simple sum, the builtin Python function `sum()` would also work fine):

```
filter_expr = functools.reduce(
    lambda x, y: x + y,
    [
        duckdb.FunctionExpression("one_if_null", duckdb.ColumnExpression(col))
        for col in df_test.columns
    ],
).alias("row_num_nulls")
```

What's missing is implementation of the `NULL` operators for `ColumnExpression`.

**Reply by cmdlineluser** [reply]
Perhaps you could open a feature request for [isNull](https://spark.apache.org/docs/3.5.0/api/python/reference/pyspark.sql/api/pyspark.sql.Column.isNull.html) / [isNotNull](https://spark.apache.org/docs/3.5.0/api/python/reference/pyspark.sql/api/pyspark.sql.Column.isNotNull.html)

> The DuckDB Spark API is currently experimental and features are still missing. We are very interested in feedback. Please report any functionality that you are missing, either through [Discord](https://discord.duckdb.org/) or on [GitHub](https://github.com/duckdb/duckdb/issues).

> All statements are translated to DuckDB’s internal plans using our [relational API](https://duckdb.org/docs/api/python/relational_api)

**Comment by ToddSmall** [comment]
I've opened a feature request to add the `IS NULL` / `IS NOT NULL` operators.

#10077
